services/CurrentTime/source/CurrentTimeService.cpp

```diff
@@ -0,0 +1,186 @@
+{
+    bool result = true;
+
+    if ((_current_time.year    < 1582) || (_current_time.year    > 9999)) {
+        result = false;
+    }
+    if ((_current_time.month   <    1) || (_current_time.month   >   12)) {
+        result = false;
+    }
+    if ((_current_time.day     <    1) || (_current_time.day     >   31)) {
+        result = false;
+    }
+    if ((_current_time.hours   <    0) || (_current_time.hours   >   23)) {
+        result = false;
+    }
+    if ((_current_time.minutes <    0) || (_current_time.minutes >   59)) {
+        result = false;
+    }
+    if ((_current_time.seconds <    0) || (_current_time.seconds >   59)) {
+        result = false;
+    }
+    if ((_current_time.weekday <    1) || (_current_time.weekday >    7)) {
+        result = false;
+    }
+
+    return result;
+}
+
+void CurrentTimeService::onDataRead(GattReadAuthCallbackParams *read_request)
+{
+    uint8_t *data = reinterpret_cast<uint8_t *>(&_current_time);
+
+    time_t local_time = get_time();
+
+    struct tm *local_time_tm = localtime(&local_time);
+
+    serialize(data, local_time_tm);
+
+    read_request->data = reinterpret_cast<uint8_t *>(&_current_time);
+    read_request->len  = CURRENT_TIME_CHAR_VALUE_SIZE;
+    read_request->authorizationReply = AUTH_CALLBACK_REPLY_SUCCESS;
+}
+
+void CurrentTimeService::onDataWritten(GattWriteAuthCallbackParams *write_request)
+{
+    const uint8_t *data = write_request->data;
+
+    struct tm remote_time_tm{};
+
+    write_request->authorizationReply = AUTH_CALLBACK_REPLY_SUCCESS;
+
+    if (!deserialize(&remote_time_tm, data)) {
+        /* The alert level is out of range
+         * */
+        write_request->authorizationReply = static_cast<GattAuthCallbackReply_t>(0xFF);
+    }
+
+    time_t remote_time = mktime(&remote_time_tm);
+
+    set_time(remote_time);
+
+    if (_current_time_handler) {
+        _current_time_handler->on_current_time_written(remote_time);
+    }
+}
+
+void CurrentTimeService::serialize(uint8_t *data, const struct tm *local_time_tm)
+{
+    *data++ = (local_time_tm->tm_year + 1900);
+    *data++ = (local_time_tm->tm_year + 1900) >> 8;
+    *data++ =  local_time_tm->tm_mon  + 1;
+    *data++ =  local_time_tm->tm_mday;
+    *data++ =  local_time_tm->tm_hour;
+    *data++ =  local_time_tm->tm_min;
+    *data++ =  local_time_tm->tm_sec;
+    /*
+     * The tm_wday field of a tm struct means days since Sunday (0-6)
+     * However, the weekday field of a CurrentTime struct means Mon-Sun (1-7)
+     * So, if tm_wday = 0, i.e. Sunday, the correct value for weekday is 7
+     * Otherwise, the fields signify the same days and no correction is needed
+     * */
+    *data++ =  local_time_tm->tm_wday == 0 ? 7 : local_time_tm ->tm_wday;
+    *data   =  0;
+}
+
+bool CurrentTimeService::deserialize(struct tm *remote_time_tm, const uint8_t *data)
+{
+    bool deserialization_is_possible;
+
+    _current_time.year    = *data | (*(data + 1) << 8);
```

Comment:
I don't expect deserialization to mutate members of the class. It means than in `onDataWritten`, it can fail and return an error but the data are committed in the service. 

I'd suggest changing `current_time_is_valid` into a member function of `CurrentTime` (`CurrentTime::is_valid`). Then add a new deserialize function from `uint8_t*` to `CurrentTime` and a conversion function from `CurrentTime` to `tm` (`struct tm toTm()`) in the `CurrentTime` class.

Reply:
I think it's fine because the _current_time member is not really used for anything - it's like a scratch space, the real time value is stored in the _time_offset which is only updated on successful deserialisation.

Reply:
It is **not** ok, the function `current_time_is_valid` is public, serialization of rejected value should not have a visible side effect to the application.

Reply:
current_time_is_valid shouldn't be public, it's not actually validating 'current' time, it's only validating what has been sent

Reply:
that member is really just scratch space, not really backing anything, the use cannot read it and the gatt server doesn't  use it